[#827] feat(operator): support generating hpa objects

wangao1236 opened the pull request:
<!--
1. Title: [#<issue>] <type>(<scope>): <subject>
   Examples:
     - "[#123] feat(operator): support xxx"
     - "[#233] fix: check null before access result in xxx"
     - "[MINOR] refactor: fix typo in variable name"
     - "[MINOR] docs: fix typo in README"
     - "[#255] test: fix flaky test NameOfTheTest"
   Reference: https://www.conventionalcommits.org/en/v1.0.0/
2. Contributor guidelines:
   https://github.com/apache/incubator-uniffle/blob/master/CONTRIBUTING.md
3. If the PR is unfinished, please mark this PR as draft.
-->

### What changes were proposed in this pull request?

In this PR, we will be adding the Autoscaler field for the first time in the CRD, enabling users to support ShuffleServer using the HPA object to achieve horizontal automatic scaling.

In the next PR, we will provide more detailed documentation to explain the format and usage examples of .spec.shuffleServer.autoscaler field.

### Why are the changes needed?

Fix: #827

### Does this PR introduce _any_ user-facing change?

If users need to use the automatic scaling feature, they need to edit the .spec.shuffleServer.autoscaler field, which is completely identical to the .spec field definition of kubernetes' HPA object.

### How was this patch tested?

Manually verified.

jerqi commented on deploy/kubernetes/operator/pkg/controller/util/kubernetes/hpa.go:
```diff
@@ -0,0 +1,104 @@
+package kubernetes
+
+import (
+	"context"
```
Could we add some tests for this file `hpa.go`?

wangao1236 replied:
Done

wangao1236 commented on deploy/kubernetes/operator/api/uniffle/v1alpha1/remoteshuffleservice_types.go:
```diff
@@ -18,6 +18,7 @@
 package v1alpha1
 
 import (
+	autoscalingv2 "k8s.io/api/autoscaling/v2beta2"
```
Yes, prior to version 1.22, only beta version is available.

wangao1236 commented on deploy/kubernetes/operator/pkg/controller/sync/shuffleserver/shuffleserver.go:
```diff
@@ -51,10 +54,13 @@ func init() {
-	return sa, sts
+	sts := GenerateSts(kubeClient, rss)
+	hpa := GenerateHPA(rss)
+	enableHPA := rss.Spec.ShuffleServer.Autoscaler.Enable
```
Ok, I have modified it.

wangao1236 commented:
> Could you also upload the test result of your HPA integrations?

I have already modified `deploy/kubernetes/integration-test/e2e/template/rss-template.yaml ` file, and enabled hpa object of rss object.

wangao1236 commented on deploy/kubernetes/operator/pkg/webhook/inspector/pod.go:
```diff
@@ -100,8 +102,33 @@ func (i *inspector) validateDeletingShuffleServer(ar *admissionv1.AdmissionRevie
+			return false
+		}
+		// In this case, we have enabled HPA and HPA triggers the StatefulSet shrink, so we should allow deletion.
+		if *sts.Spec.Replicas == hpa.Status.DesiredReplicas && *sts.Spec.Replicas < sts.Status.Replicas {
```
This may be triggered by the administrator, but the administrator is not advised to change the number of shuffle server copies after the HPA object is enabled.

wangao1236 commented on deploy/kubernetes/operator/api/uniffle/v1alpha1/remoteshuffleservice_types.go:
```diff
@@ -168,9 +169,55 @@ type CommonConfig struct {
+type RSSAutoscaler struct {
+	// Enable indicates whether we need to generate an HPA object.
+	Enable bool `json:"enable"`
+	// HPASpec allows users to configure HPA objects to achieve automatic scaling.
```
Ok, good idea.

wangao1236 commented:
### HPA Testing
Firstly, we set up a local cluster and created an RSS object within it. We enabled HPA and observed that the rss-controller not only launched the RSS component but also created an HPA object.
<img width="1790" alt="image" src="https://user-images.githubusercontent.com/12544973/233551596-78af09c6-3d7c-4ea5-bce4-770e752f7e34.png">

Next, we modified the .spec.shuffleserver.autoscaler.enable field of the RSS object to false, and reapply the /deploy/kubernetes/integration-test/e2e/rss.yaml file. We found that the HPA object has been deleted.
![image](https://user-images.githubusercontent.com/12544973/233551987-dcb077c0-e858-4cc7-b0fd-41b8d30c9a67.png)

After we re-enabled HPA, a new HPA object was created. 
![image (1)](https://user-images.githubusercontent.com/12544973/233552135-41d34f27-3089-404c-9dda-c4d88f53279f.png)

Let's take a look at the detailed information of the HPA object:
![image](https://user-images.githubusercontent.com/12544973/233552341-0098a0aa-8094-47b2-9c79-01421cb6662f.png)
![image (1)](https://user-images.githubusercontent.com/12544973/233552356-11481bcc-5706-4307-aec3-e9b03272837f.png)

We have set CPU average utilization as the dependent metric, with a threshold of 10%. Currently, it is at 3%, so there is no need for scaling up.
To trigger scaling up, we can modify the /deploy/kubernetes/integration-test/e2e/rss.yaml file, lower the threshold of HPA to 1%, and then reapply this file.
![image](https://user-images.githubusercontent.com/12544973/233552545-f0ce60f2-399e-4c65-8796-5732eea94d3b.png)

At this time, the threshold of the HPA object has also changed and the current value is greater than the threshold.
![image](https://user-images.githubusercontent.com/12544973/233552666-0043e522-f50d-43b0-a496-2dd830704f2f.png)

Two replicas can be seen created. As I have set the maximum number of replicas to three, only two replicas were added.
![image](https://user-images.githubusercontent.com/12544973/233552978-be33988f-e91c-43b7-86d5-d6a8ec396ac6.png)

Next, we would like to test scaling down.
We will modify the threshold back to 10% and then reapply the /deploy/kubernetes/integration-test/e2e/rss.yaml file.
After waiting for a while, we found that the downsizing had indeed been triggered.
The HPA Controller is generally cautious when it comes to downsizing and will not immediately downsize by a large amount.
![image (1)](https://user-images.githubusercontent.com/12544973/233553063-8ea93776-68cb-4620-947e-8429e7fa67f1.png)
![image (2)](https://user-images.githubusercontent.com/12544973/233553120-83b769bf-7211-495d-ac2c-291ed073ed1b.png)

jerqi commented:
Merged to master. thanks @advancedxy @wangao1236
